### DISCUSSION TITLE
in-graph and between-graph clustering

### DISCUSSION THREAD

**Discussion by StephDocTUM** [pyg-team/pytorch_geometric#4203] [General]
Dear all,

I want to do an in-graph and between graph clustering approach. Are there any methods already implemented especially for between-graph clustering algorithms?

Best,
Steph

**Comment by rusty1s** [comment]
Do you mean clustering of graphs rather than nodes inside a single graph? One could also use ["Deep Graph Infomax"](https://arxiv.org/abs/1809.10341) for this which is also integrated into PyG. Otherwise, please let me know of some references in which we can look into.

**Comment by StephDocTUM** [comment]
Yes, clustering of graphs. E.g. I have 1000 graphs and want to cluster the graphs into different clusters. So clustering a set of graphs:
e.g. https://web.iitd.ac.in/~bspanda/graphclustering.pdf

**Reply by rusty1s** [reply]
Yes, we indeed have not an example for this (only node-level clustering), but one should be able to customize the Deep Graph Infomax appraoch for this. Let me know if you are interested in working on this with us!

**Comment by StephDocTUM** [comment]
Yes, I would love to. Would write me an email to get started?

Best,
Stephan

**Reply by rusty1s** [reply]
No reason to go private for this. We can either discuss the roadmap here or in a separate issue. Let me know if you need any pointers on how to get started.

**Comment by StephDocTUM** [comment]
Yes sure, lets discuss the roadmap here. What do you think is a good way to start?

**Reply by StephDocTUM** [reply]
so it would be adjusting https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/nn/models/deep_graph_infomax.html

**Reply by rusty1s** [reply]
Yes, I think it should work as follow:
1. You consider pairs of randomly sampled graphs (one positive and one negative graph).
2. You compute node embeddings for both of these graphs, and a summary vector of the positive graph.
3. You pass all these information to to `DeepGraphInfomax.loss`
4. After training, the `summary` vector describes an embedding of your entire graph, learned in an unsupervised fashion.

**Reply by StephDocTUM** [reply]
Okay, I created the graphs

To calculate a node embedding of the positive and negative graph, would you recommend to use:
https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/nn/models/lightgcn.html#LightGCN.get_embedding
?

**Reply by StephDocTUM** [reply]
mhh in step 1 to 4 I cannot see where I cluster then between the graphs?

**Reply by rusty1s** [reply]
What do you mean? The loss formulation of DGI is based on maximizing mutual information. You can find more details [here](https://arxiv.org/pdf/1808.06670.pdf).

**Reply by StephDocTUM** [reply]
Yes for how DGI works.
But I would have a few more questions:
for step 1.) Do you mean a lot of different randomly sampled graphs (and split each graph into a positive and negative one)?
4.) The result of step 4 is the representation of ONE graph as summary vector?
Then I would have to do my clustering in the next step with all these summary vectors?

**Reply by rusty1s** [reply]
1) I mean pairs of randomly sampled graphs (at best loaded in a batch-wise fashion), e.g. :
```python
for pos_batch, neg_batch in zip(pos_loader, neg_loader):
    pos_out = model(pos_batch)
    neg_out = model(neg_batch)
    pos_summary = global_mean_pool(pos_out, pos_batch.batch)
    ...
```
4) Yes, the clustering is performed in a post-processing step.

**Reply by StephDocTUM** [reply]
For the global_mean_pool, I run:

pos_summary = global_mean_pool(embedding_pos, data_pos.x)

where embedding_pos is the embedding of the positiv graph and data_pos.x are the nodes of the initial positive graph.

Yields in:
RuntimeError: The expanded size of the tensor (243) must match the existing size (128) at non-singleton dimension 1.  Target sizes: [6668, 243].  Tensor sizes: [6668, 128]

 data_pos.x should be a vector?

**Reply by rusty1s** [reply]
Yes, the `index` needs to be a vector and of type `torch.long`. That is, it denotes how we want to aggregate `embedding_pos`. See [here](https://github.com/rusty1s/pytorch_scatter) for a visualization.

**Reply by StephDocTUM** [reply]
so for `pos_summary` I receive `[2,128]` dimension, because I set the latent dimension to 128.
Then I created an object of DeepGraphInfomax only with the information of dim = 128 and not using:

```
encoder (Module): The encoder module :math:`\mathcal{E}`.
summary (callable): The readout function :math:`\mathcal{R}`.
corruption (callable): The corruption function :math:`\mathcal{C}`.
```

Because, I only want to use .loss and .discriminate form the DGI class. But in .discriminate I multiply:
`        value = torch.matmul(z, torch.matmul(self.weight, summary)) 
`
where `self.weight` is of dim `[128,128]` (from the init function `self.weight = Parameter(torch.Tensor(hidden_channels, hidden_channels))` and summary is of `[2,128]` yielding in `RuntimeError: mat1 and mat2 shapes cannot be multiplied (128x128 and 2x128)`

**Reply by rusty1s** [reply]
Currently, our "Deep Graph Infomax" implementation expects a vector as `summary`. If you want to pass in a matrix, you will need to `transpose` it first:
```python
value = torch.matmul(z, torch.matmul(self.weight, summary.t())) 
```
Let me know if you have interest in fixing this in our implementation!

**Reply by StephDocTUM** [reply]
When I am passing it transposed it works and returns me the mutual information maximization objective? So it should work without fixing the implementation?
When I only call DeepGraphInfomax.loss, where is my final summary vector of the positive graph updated, which I use in the end for clustering? 
In DeepGraphInfomax the updating takes place in the forward method (I used as encoder Node2vec).

**Reply by rusty1s** [reply]
Yes, italso works when you first transpose prior to calling this function, but this is a bit unexpected. I am still in favor of fixing it directly inside `DeepGraphInfomax`, e.g., via:
```python
summary = summary.t() if summary.dim() > 1 else summary
```

**Reply by rusty1s** [reply]
Fixed in https://github.com/pyg-team/pytorch_geometric/pull/4402.

**Reply by StephDocTUM** [reply]
yes and When I only call DeepGraphInfomax.loss, where is my final summary vector of the positive graph updated, which I use in the end for clustering?
In DeepGraphInfomax the updating takes place in the forward method (I used as encoder Node2vec)?

**Reply by rusty1s** [reply]
I am not sure I understand. IMO, you can directly use the `summary` output of `model.forward` to perform clustering, right?

**Reply by StephDocTUM** [reply]
Yes.
One question for the summary function = readout function in DGI. I have to specify this function to be able to run the foward() function in DGI.
Where is this summary function defined and how does it look like?

**Reply by rusty1s** [reply]
The `summary` function is defined here: https://github.com/pyg-team/pytorch_geometric/blob/master/examples/infomax_transductive.py#L33

**Comment by StephDocTUM** [comment]
What about:

```
def corruption(x, edge_index):
    return x[torch.randperm(x.size(0))], edge_index

def readout(z, *args, **kwargs):
    return torch.sigmoid(z.mean(dim=0))

def train_one_epoch(model,data):
    data = data.to(DEVICE)
    x = data.x.to(DEVICE)
    edge_index = data.edge_index.to(DEVICE)
    model.train() #wo wird das ausgeführt?
    optimizer.zero_grad()
    pos_z, neg_z, summary = model(x.float(), edge_index) ###where does model go with data.x,..
    loss = model.loss(pos_z, neg_z, summary)
    loss.backward()
    optimizer.step()
    return loss.item()



## SETTING UP THE MODEL
feature_dim = data.x.shape[1]
model = DeepGraphInfomax(
        hidden_channels=EMB_DIM,
        encoder= Encoder(feature_dim,EMB_DIM),
        summary=readout,
        corruption=corruption).to(DEVICE)

# Embedding space before training the model
model.eval()
x = data.x.to(DEVICE)
edge_index = data.edge_index.to(DEVICE)
untrained_out = model(x.float(), edge_index)

# Training
optimizer = torch.optim.Adam(model.parameters(), lr=LR)

for epoch in range(EPOCH):

    loss = train_one_epoch(model,data)
    print('Epoch: {:03d}, Loss: {:.4f}'.format(epoch, loss))

# Embedding space after training
model.eval()
trained_out = model(x.float(), edge_index)
```

then

trained_out[2] should give me the vector of the embedding of my entire graph right?

**Reply by rusty1s** [reply]
Yes, that is correct.

**Reply by StephDocTUM** [reply]
So, this can be used for in-between graph clustering right?

**Reply by rusty1s** [reply]
I think so. You can use examples from the training set as corrupted examples.
